simulations/vip-178/vip-178/simulations.ts

```diff
@@ -0,0 +1,320 @@
+
+        before(async () => {
+          vToken = await ethers.getContractAt(VTOKEN_ABI, vTokenAddress);
+        });
+
+        it(`should have name = "${name}"`, async () => {
+          expect(await vToken.name()).to.equal(name);
+        });
+
+        it(`should have symbol = "${symbol}"`, async () => {
+          expect(await vToken.symbol()).to.equal(symbol);
+        });
+
+        it(`should have ${decimals.toString()} decimals`, async () => {
+          expect(await vToken.decimals()).to.equal(decimals);
+        });
+
+        it(`should have underlying = "${underlying}"`, async () => {
+          expect(await vToken.underlying()).to.equal(underlying);
+        });
+
+        it(`should have initial exchange rate of ${exchangeRate.toString()}`, async () => {
+          expect(await vToken.exchangeRateStored()).to.equal(exchangeRate);
+        });
+
+        it("should have the correct Comptroller", async () => {
+          expect(await vToken.exchangeRateStored()).to.equal(exchangeRate);
+        });
+      });
+    };
+
+    checkVToken(VagEUR_Stablecoins, {
+      name: "Venus agEUR (Stablecoins)",
+      symbol: "vagEUR_Stablecoins",
+      decimals: 8,
+      underlying: agEUR,
+      exchangeRate: parseUnits("1", 28),
+    });
+  });
+
+  testVip("VIP-178 Add agEUR Market", vip178(24 * 60 * 60 * 3), {
+    callbackAfterExecution: async (txResponse: TransactionResponse) => {
+      await expectEvents(
+        txResponse,
+        [COMPTROLLER_ABI, POOL_REGISTRY_ABI, ERC20_ABI, REWARD_DISTRIBUTOR_ABI],
+        [
+          "Approval",
+          "MarketAdded",
+          "NewRewardsDistributor",
+          "RewardTokenSupplySpeedUpdated",
+          "RewardTokenBorrowSpeedUpdated",
+          "OwnershipTransferred",
+        ],
+        [6, 1, 1, 0, 1, 3],
+      );
+    },
+  });
+
+  describe("Post-VIP state", () => {
+    describe("PoolRegistry state", () => {
+      it("should register pool's vTokens in Comptroller", async () => {
+        const vTokens = await comptroller.getAllMarkets();
+        expect(vTokens).to.have.lengthOf(4);
+        expect(vTokens).to.include(VagEUR_Stablecoins);
+      });
+
+      it("should register in PoolRegistry", async () => {
+        const vToken = await poolRegistry.getVTokenForAsset(STABLECOIN_COMPTROLLER, agEUR);
+        expect(vToken).to.equal(VagEUR_Stablecoins);
+      });
+    });
+
+    describe("Ownership", () => {
+      it("should transfer ownership to Timelock", async () => {
+        expect(await vagEUR.owner()).to.equal(NORMAL_TIMELOCK);
+        expect(await rewardsDistributor.owner()).to.equal(NORMAL_TIMELOCK);
+      });
+    });
+
+    describe("Initial supply", () => {
+      it(`should mint 9,000 vagEUR to ${VTOKEN_RECEIVER_agEUR}`, async () => {
+        expect(await vagEUR.balanceOf(VTOKEN_RECEIVER_agEUR)).to.equal(parseUnits("9000", 8));
+      });
+    });
+
+    describe("Market and risk parameters", () => {
+      const checkInterestRate = (
```

Comment:
We could move this check function to this folder, to reuse it in the future: https://github.com/VenusProtocol/vips/blob/develop/src/vip-framework/checks